presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -34,23 +37,22 @@ private Util()
 
     /**
      * Prune the list of available inputs to those required by the given expressions.
-     *
+     * <p>
      * If all inputs are used, return Optional.empty() to indicate that no pruning is necessary.
      */
     public static Optional<List<Symbol>> pruneInputs(Collection<Symbol> availableInputs, Collection<Expression> expressions)
     {
-        Set<Symbol> available = new HashSet<>(availableInputs);
-        Set<Symbol> required = DependencyExtractor.extractUnique(expressions);
-
         // we need to compute the intersection in case some dependencies are symbols from
         // the outer scope (i.e., correlated queries)
-        Set<Symbol> used = Sets.intersection(required, available);
-        if (used.size() == available.size()) {
+        List<Symbol> prunedInputs = availableInputs.stream()
+                .filter(DependencyExtractor.extractUnique(expressions)::contains)
```

Comment:
I think streams are not that smart to avoid calling `extractUnique` for every symbol. You may want to retrieve it once and pass to the stream-processor.